### PR TITLE
Support command line parameters in benchmarks

### PR DESCRIPTION
**Pull request by cz4rs**
Support command line parameters matching the behavior of original performance tests as closely as possible.

Sample output (console):
```
./build/perf_test/KokkosKernels_PerformanceTest_Benchmark --benchmark_format=console --benchmark_counters_tabular=true --m 99 --n 3 --repeat 2 --benchmark_filter=Blas1_dot_mv --use_openmp 1
```
```
Running ./build/perf_test/KokkosKernels_PerformanceTest_Benchmark
(...)
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations Avg DOT FLOP/s: Avg DOT time (s):     OpenMP  Params::m  Params::n Params::repeat Params::use_cuda Params::use_hip Params::use_openmp Params::use_sycl Params::use_threads
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Blas1_dot_mv<double>/manual_time      0.000 s         0.000 s         44468         85.443M            6.952u          1         99          3              2                0               0                  1                0                   0
```

Sample output (json):
```json
{
  "context": {
(...)
    "Default Device": "N6Kokkos6OpenMPE",
    "GIT_BRANCH": "benchmark-command-line-params",
    "GIT_CLEAN_STATUS": "CLEAN",
    "GIT_COMMIT_DATE": "2023-03-30T15:57:56+02:00",
    "GIT_COMMIT_DESCRIPTION": "Parse command line parameters",
    "GIT_COMMIT_HASH": "f58c6469f",
    "GOOGLE_BENCHMARK_VERSION": "1.6.2",
(...)
  },
  "benchmarks": [
    {
      "name": "Blas1_dot_mv<double>/manual_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "Blas1_dot_mv<double>/manual_time",
      "run_type": "iteration",
      "repetitions": 1,
      "repetition_index": 0,
      "threads": 1,
      "iterations": 44285,
      "real_time": 1.5161024026193994e-05,
      "cpu_time": 2.3002449520153550e-05,
      "time_unit": "s",
      "Avg DOT FLOP/s:": 8.5566119273984447e+07,
      "Avg DOT time (s):": 6.9419999999999998e-06,
      "OpenMP": 1.0000000000000000e+00,
      "Params::m": 9.9000000000000000e+01,
      "Params::n": 3.0000000000000000e+00,
      "Params::repeat": 2.0000000000000000e+00,
      "Params::use_cuda": 1.0000000000000000e+00,
      "Params::use_hip": 0.0000000000000000e+00,
      "Params::use_openmp": 1.0000000000000000e+00,
      "Params::use_sycl": 0.0000000000000000e+00,
      "Params::use_threads": 0.0000000000000000e+00
    }
  ]
}
```

Error when requesting unavailable `ExecSpace`:
```
Running ./build/perf_test/KokkosKernels_PerformanceTest_Benchmark
(...)
---------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations
---------------------------------------------------------------------------
Blas1_dot_mv<double>/manual_time ERROR OCCURRED: 'CUDA requested, but not available.'
```

### CODE REVIEW

File: perf_test/Benchmark_Context.hpp [hunk 1]
```diff
@@ -95,6 +95,21 @@ inline void add_benchmark_context(bool verbose = false) {
   add_version_info();
 }
 
+struct Params {
+  static std::map<std::string, int> params_;
+
+  static void parse_inputs(int argc, char** argv) {
+    for (int i = 1; i < argc; i += 2) {
+      params_[argv[i]] = std::atoi(argv[i + 1]);
```
Comment (lucbv) [inline perf_test/Benchmark_Context.hpp:103]: How do you know every parameter will always be an `int` and not a `bool`, a string or a floating point values?
Maybe this should be called `parse_int_inputs` and additional version should be added for other types?

File: perf_test/Benchmark_Context.hpp [hunk 2]
```diff
@@ -95,6 +95,21 @@ inline void add_benchmark_context(bool verbose = false) {
...
+    }
+  }
+
+  static int get_param_or_default(std::string param, int default_value) {
```
Comment (lucbv) [inline perf_test/Benchmark_Context.hpp:107]: Same as above

File: perf_test/Benchmark_Context.hpp [hunk 3]
```diff
@@ -95,6 +95,21 @@ inline void add_benchmark_context(bool verbose = false) {
   add_version_info();
 }
 
+struct Params {
+  static std::map<std::string, int> params_;
+
+  static void parse_inputs(int argc, char** argv) {
+    for (int i = 1; i < argc; i += 2) {
```
Comment (lucbv) [inline perf_test/Benchmark_Context.hpp:102]: This assumes something about all the arguments being of the form `--option value` and none being of the form `--use_option`

**Review by lucbv — CHANGES_REQUESTED** [review]
This needs to be discussed, I think this design is bad and will lead to a lot of trouble down the road when the benchmarks are more complex than Blas.

**Comment by cz4rs** [comment]
> This needs to be discussed, I think this design is bad and will lead to a lot of trouble down the road when the benchmarks are more complex than Blas.

I agree on the need for more general approach. This is matching current implementation for Blas_dot_mv:
https://github.com/kokkos/kokkos-kernels/blob/b079a4e2dd76de85e9024cb5c447cdb2b3c1293b/perf_test/blas/blas1/KokkosBlas_dot_mv_perf_test.cpp#L53-L85

The one that is probably most complex amongst Blas tests is `KokkosBlas3_perf_test.cpp`:
https://github.com/kokkos/kokkos-kernels/blob/b079a4e2dd76de85e9024cb5c447cdb2b3c1293b/perf_test/blas/blas3/KokkosBlas3_perf_test.cpp#L25-L28
https://github.com/kokkos/kokkos-kernels/blob/b079a4e2dd76de85e9024cb5c447cdb2b3c1293b/perf_test/blas/blas3/KokkosBlas3_perf_test.cpp#L276-L298

**Comment by kokkos-devops-admin** [comment]
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO INSPECTION HAS BEEN PERFORMED ON THIS PULL REQUEST! - This PR must be inspected by setting label 'AT: PRE-TEST INSPECTED'.

**Comment by cz4rs** [comment]
Closing as the preferred approach was merged with #1754.

**Pull request closed without merging**